app/plugin_install.go

```diff
@@ -333,21 +339,81 @@ func (a *App) removePlugin(id string) *model.AppError {
 	if err = a.removeSignature(id); err != nil {
 		mlog.Error("Can't remove signature", mlog.Err(err))
 	}
-
-	a.notifyClusterPluginEvent(
-		model.CLUSTER_EVENT_REMOVE_PLUGIN,
-		model.PluginEventData{
-			Id: id,
-		},
-	)
+	canaryTag := *a.Config().PluginSettings.CanaryTag
+	if canaryTag == "" {
+		a.notifyClusterPluginEvent(
+			model.CLUSTER_EVENT_REMOVE_PLUGIN,
+			model.PluginEventData{
+				Id: id,
+			},
+		)
+	}
 
 	if err := a.notifyPluginStatusesChanged(); err != nil {
 		mlog.Error("Failed to notify plugin status changed", mlog.Err(err))
 	}
 
+	if canaryTag == "" {
+		a.removeAllPlugins(id)
+	} else {
+		a.installPluginWithNoCanaryTag(id, pluginEnabled)
+	}
+
 	return nil
 }
 
+func (a *App) installPluginWithNoCanaryTag(id string, pluginEnabled bool) {
```

Comment:
Would it be feasible to reuse the existing `installPlugin`, but accept the canary tag and pass it as empty? I'm worried about the presumed duplication of code here and the potential for the canary path to diverge.

Reply:
There is no duplication here. This method is called when we remove plugin with canary tag(basically plugin we are testing) and we want to fall back on the plugin with no canary tag("stable" implementation of a plugin). It searches for the plugin and signature file, installs and enables if it was previously enabled.

Reply:
What we might want to change here is logging in the Warn level and not Error level when there is no fallback.